File: 2022/04/04.py
sections = []

with open('input.txt', 'r') as file:
    section = []
    for line in file:
        l = line.strip()
        section.append(int(l.split('-')[0]))
        section.append(int(l.split(',')[0].split('-')[1]))
        section.append(int(l.split(',')[1].split('-')[0]))
        section.append(int(l.split('-')[2]))
        
        sections.append(section.copy())
        section.clear()

#print(sections)

def puzzle1(sections):
    count = 0
    for section in sections:
        if section[0] >= section[2] and section[1] <= section[3]:
            count += 1
        elif section[2] >= section[0] and section[3] <= section[1]:
            count += 1

    return count

def puzzle2(sections):
    count = 0

    for section in sections:
        if section[1] == section[2] or section[3] == section[0]:
            count += 1
        elif section[1] <= section[3] and section[1] >= section[2]:
            count += 1
        elif section[3] <= section[1] and section[3] >= section[0]:
            # 4-8,2-6
            count += 1
        elif section[0] >= section[2] and section[1] <= section[3]:
            count += 1
        elif section[2] >= section[0] and section[3] <= section[1]:
            count += 1

    return count


print(puzzle1(sections))
print(puzzle2(sections))

File: 2022/04/04.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

void main() async {
  File file = File('input.txt');
  Stream<String> lines =
      file.openRead().transform(utf8.decoder).transform(LineSplitter());

  List sections = [];

  try {
    List section = [];
    await for (var line in lines) {
      section.add(int.parse(line.split('-')[0]));
      section.add(int.parse(line.split(',')[0].split('-')[1]));
      section.add(int.parse(line.split(',')[1].split('-')[0]));
      section.add(int.parse(line.split('-')[2]));

      sections.add(List.from(section));
      section.clear();
    }
  } catch (error) {
    print('Error $error');
  }

  int count1 = 0;
  int count2 = 0;

  sections.forEach((s) {
    if ((s[0] >= s[2] && s[1] <= s[3]) || (s[2] >= s[0] && s[3] <= s[1]))
      count1++;
    if (!(s[1] < s[2] || s[3] < s[0])) count2++;
  });

  print(count1);
  print(count2);
}
